code/game/machinery/telecomms/computers/telemonitor.dm

```diff
@@ -1,126 +1,156 @@
-/*
-	Telecomms monitor tracks the overall trafficing of a telecommunications network
-	and displays a heirarchy of linked machines.
-*/
-
 #define MAIN_VIEW 0
 #define MACHINE_VIEW 1
 #define MAX_NETWORK_ID_LENGTH 15
 
 /obj/machinery/computer/telecomms/monitor
 	name = "telecommunications monitoring console"
-	icon_screen = "comm_monitor"
 	desc = "Monitors the details of the telecommunications network it's synced with."
 
-	/// Current screen the user is viewing
-	var/screen = MAIN_VIEW
-	/// Weakrefs of the machines located by the computer
-	var/list/machine_list = list()
+	circuit = /obj/item/circuitboard/computer/comm_monitor
+
+	icon_screen = "comm_monitor"
+
 	/// Weakref of the currently selected tcomms machine
-	var/datum/weakref/selected_machine_ref
+	var/datum/weakref/selected_machine_ref = null
+	/// Weakrefs of the machines located by the computer
+	var/list/datum/weakref/machine_refs
+
+	/// Currently displayed "tab"
+	var/screen = MAIN_VIEW
 	/// The network to probe
 	var/network = "NULL"
 	/// Error message to show
-	var/error_message = ""
-	circuit = /obj/item/circuitboard/computer/comm_monitor
+	var/status_message = null
 
-/obj/machinery/computer/telecomms/monitor/ui_data(mob/user)
-	var/list/data = list(
-		"screen" = screen,
-		"network" = network,
-		"error_message" = error_message,
+/obj/machinery/computer/telecomms/monitor/Initialize(mapload, obj/item/circuitboard/C)
+	. = ..()
+	LAZYINITLIST(machine_refs)
+
+/obj/machinery/computer/telecomms/monitor/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/telecomms)
 	)
 
+/obj/machinery/computer/telecomms/monitor/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "TelecommsMonitor")
+		ui.open()
+
+/obj/machinery/computer/telecomms/monitor/ui_data(mob/user)
+	var/list/data = list()
+
+	data["screen"] = screen
+	data["statusMessage"] = status_message
+
 	switch(screen)
-	  	// --- Main Menu ---
 		if(MAIN_VIEW)
-			var/list/found_machinery = list()
-			for(var/datum/weakref/tcomms_ref in machine_list)
-				var/obj/machinery/telecomms/telecomms = tcomms_ref.resolve()
-				if(!telecomms)
-					machine_list -= tcomms_ref
+			data["network"] = network
+
+			data["machines"] = list()
+			for(var/datum/weakref/machine_ref in machine_refs)
+				var/obj/machinery/telecomms/machine = machine_ref.resolve()
+				if(isnull(machine))
+					machine_refs -= machine_ref
 					continue
-				found_machinery += list(list("ref" = REF(telecomms), "name" = telecomms.name, "id" = telecomms.id))
-			data["machinery"] = found_machinery
-	  	// --- Viewing Machine ---
+
+				data["machines"] += list(list(
+					"id" = machine.id,
+					"name" = machine.name,
+					"icon" = initial(machine.icon_state)
```

Comment:
```suggestion
					"icon" = initial(machine.icon_state),
```
Should this be using ``base_icon_state`` instead?

Reply:
No, because I use reference to an initial icon during sprite sheet generation. At least I think so.